zebra-chain/src/transparent/address.rs

```diff
@@ -101,6 +116,43 @@ impl From<Script> for Address {
     }
 }
 
+/// Returns the address that this Script contains, if any.
+/// Ported from https://github.com/zcash/librustzcash/blob/master/zcash_primitives/src/legacy.rs#L43
+impl From<ScriptForNetwork> for Address {
+    fn from(script_network: ScriptForNetwork) -> Self {
+        let script = script_network.script;
+        let network = script_network.network;
+
+        if script.0.len() == 25
+            && script.0[0..3] == [OpCode::Dup as u8, OpCode::Hash160 as u8, 0x14]
+            && script.0[23..25] == [OpCode::EqualVerify as u8, OpCode::CheckSig as u8]
+        {
+            let mut script_hash = [0; 20];
+            script_hash.copy_from_slice(&script.0[3..23]);
+
+            Address::PayToScriptHash {
+                network,
+                script_hash,
+            }
+        } else if script.0.len() == 23
+            && script.0[0..2] == [OpCode::Hash160 as u8, 0x14]
+            && script.0[22] == OpCode::Equal as u8
+        {
+            let mut script_hash = [0; 20];
+            script_hash.copy_from_slice(&script.0[2..22]);
+            Address::PayToScriptHash {
+                network,
+                script_hash,
+            }
+        } else {
+            Address::PayToScriptHash {
+                network,
+                script_hash: Self::hash_payload(&script.0[..]),
+            }
```

Comment:
Why do we do `hash_payload` (and the Sha256 and Ripemd160 hashes) in the final case, but not the first two cases?

@hdevalence or @dconnolly, I'm going to need you to review this script code, because I'm not familiar with Zcash or Bitcoin script.

Reply:
The original function from librustzcash https://github.com/zcash/librustzcash/blob/master/zcash_primitives/src/legacy.rs#L43 returns `Option<Address>` and the third case returns `None`. The `hash_payload` was added by me basically because i had plans to replace `From<Script> for Address` with this one at some point and use that as default however i am not totally sure.

Reply:
I checked https://github.com/ZcashFoundation/zebra/pull/1191 (great work) and the code here(or parts of it) will need to be placed in `to_address()`. This is inline with what i posted above where 2 new cases are added and the default is what we have now.

Reply:
Rebased to use https://github.com/ZcashFoundation/zebra/pull/1191

Reply:
Ping @hdevalence or @dconnolly.
I'm going to need you to review this script code, because I'm not familiar with Zcash or Bitcoin script.

Why do we do `hash_payload` (and the Sha256 and Ripemd160 hashes) in the final case, but not the first two cases?